### DISCUSSION TITLE
SQLAlchemy Queuepool limit exhausted when using FastAPI Depends for DB session connections

### DISCUSSION THREAD

**Discussion by Laksh47** [fastapi/fastapi#10450] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List

from fastapi import Depends, FastAPI
from pydantic import BaseModel
from sqlalchemy import Column, Integer, String
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

from contextlib import contextmanager

SQLALCHEMY_DATABASE_URL = "postgresql://postgres:1234@127.0.0.1:5432/postgres"

engine = create_engine(SQLALCHEMY_DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, index=True)


class UserBase(BaseModel):
    class Config:
        orm_mode = True

    email: str


def get_users(session: Session):
    return session.query(User).all()


Base.metadata.create_all(bind=engine)


# Dependency
def get_session():
    session = SessionLocal()
    try:
        yield session
    finally:
        session.close()

@contextmanager
def get_session_with_ctx_manager():
    session = SessionLocal()
    try:
        yield session
    finally:
        session.close()


app = FastAPI()


@app.get("/users", response_model=List[UserBase])
def read_users(session: Session = Depends(get_session)):
    users = get_users(session)
    return users


@app.get("/users_fixed", response_model=List[UserBase])
def read_users_ctx_mgr():
    with get_session_with_ctx_manager() as session:
        users = get_users(session)
    return users
```

## Issue
- Facing SQLAlchemy QueuePool limit exceeded issue while using DB as a dependency as recommended by [FastAPI docs here](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-dependency) using `Depends`
   - Underlying issue seems to be the connections opened are not properly closed, leading to connections exhausted over time when we try to open new connections
- Potential workaround seems to be using a context manager for DB sessions

## Replication steps
- Create a below simple FastAPI app (inspired from a [previous example](https://github.com/andersonrocha0/fastApiAndSqlAlchemySyncVsAsync/tree/main) by @andersonrocha0)
```python
from typing import List

from fastapi import Depends, FastAPI
from pydantic import BaseModel
from sqlalchemy import Column, Integer, String
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

from contextlib import contextmanager

SQLALCHEMY_DATABASE_URL = "postgresql://postgres:1234@127.0.0.1:5432/postgres"

engine = create_engine(SQLALCHEMY_DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()

class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, index=True)


class UserBase(BaseModel):
    class Config:
        orm_mode = True

    email: str

def get_users(session: Session):
    return session.query(User).all()

Base.metadata.create_all(bind=engine)

# Dependency
def get_session():
    session = SessionLocal()
    try:
        yield session
    finally:
        session.close()

@contextmanager
def get_session_with_ctx_manager():
    session = SessionLocal()
    try:
        yield session
    finally:
        session.close()

app = FastAPI()

@app.get("/users", response_model=List[UserBase])
def read_users(session: Session = Depends(get_session)):
    users = get_users(session)
    return users

@app.get("/users_fixed", response_model=List[UserBase])
def read_users_ctx_mgr():
    with get_session_with_ctx_manager() as session:
        users = get_users(session)
    return users
```
- Run a local postgres docker container with username `postgres`, password `1234`
```docker
(3.11.0/envs/test) sync % docker ps
CONTAINER ID   IMAGE                  COMMAND                  CREATED       STATUS       PORTS                    NAMES
6fac75534a55   postgres:15.0-alpine   "docker-entrypoint.s…"   4 hours ago   Up 4 hours   0.0.0.0:5432->5432/tcp   fastapiandsqlalchemysyncvsasync-db-sync-async-test-1
(3.11.0/envs/test) sync % 
```
- Run the FastAPI app, ex: `uvicorn sync:app --reload --port 8000`
- Use [wrk](https://github.com/wg/wrk/tree/master) to simulate multiple users/concurrent requests.
- Locally running `wrk -t12 -c96 -d10s http://localhost:8000/users` will return `QueuePool limit of size 5 overflow 10 reached` error
- Locally running `wrk -t12 -c96 -d10s http://localhost:8000/users_fixed` will not throw the same error
- Suggests that the FastAPI recommended way has a bug in terms of properly closing connections

## More context

- [Background](https://docs.sqlalchemy.org/en/13/errors.html#queuepool-limit-of-size-x-overflow-y-reached-connection-timed-out-timeout-z) on the Queuepool limit issue
```shell
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)
```
- Previous discussions [around the same issue](https://github.com/tiangolo/full-stack-fastapi-postgresql/issues/104) 
   - Possible workaround suggested in the [same thread here](https://github.com/tiangolo/full-stack-fastapi-postgresql/issues/104#issuecomment-586466934)

## requirements file for accurate environment info
```python
annotated-types==0.6.0
    # via pydantic
anyio==3.7.1
    # via
    #   fastapi
    #   starlette
black==23.9.1
    # via -r requirements.in
click==8.1.7
    # via
    #   black
    #   uvicorn
fastapi==0.103.2
    # via -r requirements.in
h11==0.14.0
    # via uvicorn
idna==3.4
    # via anyio
mypy-extensions==1.0.0
    # via black
packaging==23.2
    # via black
pathspec==0.11.2
    # via black
platformdirs==3.11.0
    # via black
psycopg2-binary==2.9.9
    # via -r requirements.in
pydantic==2.4.2
    # via fastapi
pydantic-core==2.10.1
    # via pydantic
sniffio==1.3.0
    # via anyio
sqlalchemy==2.0.22
    # via -r requirements.in
starlette==0.27.0
    # via fastapi
typing-extensions==4.8.0
    # via
    #   fastapi
    #   pydantic
    #   pydantic-core
    #   sqlalchemy
uvicorn==0.23.2
    # via -r requirements.in
```

### Operating System

macOS

### Operating System Details

- using a `pyenv` virtual env for this demo app (python version: `3.11.0`)

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.11.0

### Postgres container version

postgres:15.0-alpine

## Issue endpoint output
- Running the endpoint that exhausts the connection pool
```shell
(3.11.0/envs/test) sync % wrk -t12 -c96 -d10s http://localhost:8000/users      
Running 10s test @ http://localhost:8000/users
  12 threads and 96 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    32.29ms  347.19us  32.53ms  100.00%
    Req/Sec    20.00      0.00    20.00    100.00%
  2 requests in 10.03s, 252.00B read
Requests/sec:      0.20
Transfer/sec:      25.13B
```
- Error thrown
```shell
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/applications.py", line 292, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/lib/python3.11/contextlib.py", line 222, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/concurrency.py", line 36, in contextmanager_in_threadpool
    raise e
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)
```

## Fix/workaround endpoint output
- - Running the endpoint that doesn't exhaust the connection pool
```shell
(3.11.0/envs/test) sync % wrk -t12 -c96 -d10s http://localhost:8000/users_fixed
Running 10s test @ http://localhost:8000/users_fixed
  12 threads and 96 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    78.74ms   17.46ms 145.65ms   68.24%
    Req/Sec   101.48     23.44   171.00     74.67%
  12212 requests in 10.07s, 1.47MB read
Requests/sec:   1212.55
Transfer/sec:    149.20KB
(3.11.0/envs/test)
```
- **No errors**

**Comment by Laksh47** [comment]
# Update

- Another recommended way of creating a DB session using middleware like [it is suggested here in the docs](https://fastapi.tiangolo.com/tutorial/sql-databases/#alternative-db-session-with-middleware) is throwing the same issue!

## Sample codeblock:
```python
from typing import List

from fastapi import Depends, FastAPI, Request, Response
from pydantic import BaseModel
from sqlalchemy import Column, Integer, String
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

SQLALCHEMY_DATABASE_URL = "postgresql://postgres:1234@127.0.0.1:5432/postgres"

engine = create_engine(SQLALCHEMY_DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, index=True)


class UserBase(BaseModel):
    class Config:
        orm_mode = True

    email: str


def get_users(session: Session):
    return session.query(User).all()


Base.metadata.create_all(bind=engine)


# Dependency
def get_session(request: Request):
    return request.state.db


app = FastAPI()


@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    response = Response("Internal server error", status_code=500)
    try:
        request.state.db = SessionLocal()
        response = await call_next(request)
    finally:
        request.state.db.close()
    return response


@app.get("/users", response_model=List[UserBase])
def read_users(session: Session = Depends(get_session)):
    users = get_users(session)
    return users

```

## Sample run
```shell
(3.11.0/envs/test) sync % wrk -t1 -c60 -d10s http://localhost:8000/users
Running 10s test @ http://localhost:8000/users
  1 threads and 60 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.00us    0.00us   0.00us     nan%
    Req/Sec     0.00      0.00     0.00       nan%
  0 requests in 10.10s, 0.00B read
Requests/sec:      0.00
Transfer/sec:       0.00B
(3.11.0/envs/test) sync %
```

## Error
```shell
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/streams/memory.py", line 98, in receive
    return self.receive_nowait()
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/streams/memory.py", line 93, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/base.py", line 78, in call_next
    message = await recv_stream.receive()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/streams/memory.py", line 118, in receive
    raise EndOfStream
anyio.EndOfStream

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/applications.py", line 292, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/base.py", line 108, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/Desktop/repos/fastApiAndSqlAlchemySyncVsAsync/sync/sync.py", line 52, in db_session_middleware
    response = await call_next(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/base.py", line 84, in call_next
    raise app_exc
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/base.py", line 70, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/routing.py", line 273, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/fastapi/routing.py", line 192, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/Desktop/repos/fastApiAndSqlAlchemySyncVsAsync/sync/sync.py", line 60, in read_users
    users = get_users(session)
            ^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/Desktop/repos/fastApiAndSqlAlchemySyncVsAsync/sync/sync.py", line 33, in get_users
    return session.query(User).all()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/query.py", line 2693, in all
    return self._iter().all()  # type: ignore
           ^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/query.py", line 2847, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
                                                  ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2306, in execute
    return self._execute_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2178, in _execute_internal
    conn = self._connection_for_bind(bind)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2045, in _connection_for_bind
    return trans._connection_for_bind(engine, execution_options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 2, in _connection_for_bind
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1143, in _connection_for_bind
    conn = bind.connect()
           ^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 3268, in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 3292, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/pool/base.py", line 1269, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
  File "/Users/Laksh.Arumugam/.pyenv/versions/3.11.0/envs/test/lib/python3.11/site-packages/sqlalchemy/pool/impl.py", line 158, in _do_get
    raise exc.TimeoutError(
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)
```

**Reply by lricardo** [reply, marked as answer]
Hi! The problem is not related with FastAPI itself but rather the way we need to configure SQLAlchemy.

I made a post explaining that [here](https://lricardo.space/posts/asynchronous-sqlalchemy/)

TL;DR is:
* Don't use `QueuePool`; use `AsyncAdaptedQueuePool` (this is undocumented)
* Don't use `sessionmaker`; use `async_sessionmaker`

QueuePool doesn't support async operations (see the implementation)
